jwt_proxy/api.py

```diff
@@ -2,12 +2,41 @@
 import jwt
 import requests
 import json
+import re
 
 from jwt_proxy.audit import audit_HAPI_change
 
 blueprint = Blueprint('auth', __name__)
 SUPPORTED_METHODS = ('GET', 'POST', 'PUT', 'DELETE', 'OPTIONS')
 
+# TODO: to be pulled into its own module and loaded per config
+def scope_filter(req, token):
+    # Check path
+    resource_pattern = rf"(Patient|DocumentReference)$"
+    if re.search(resource_pattern, req.path) is None:
+        return False
+    
+    user_id = token.get("sub")
+    identifier_pattern = rf"(https(:|%3[Aa])(\/|%2[Ff]){2}keycloak\.ltt\.cirg\.uw\.edu(%7[Cc]|\|))?{user_id}"
```

Comment:
this is rather hardcoded to one keycloak install.  it would be nice to look to a config value, similar to `UPSTREAM_SERVER`.

any reason to use `re` for URL parsing, vs. say `urlparse` from `urllib.parse` ?

Reply:
Thanks Paul, good points -

This pattern just matches `identifier`-like query param values like `https://keycloak.ltt.cirg.uw.edu|keycloak-user-id`
The code system `https://keycloak.ltt.cirg.uw.edu` is a dummy system url that cPRO adds to the LTT patient resources to identify keycloak user ids, and would certainly better live in config to support other codings/identifiers/systems.

I wasn't aware of `urlparse`, so thanks for the recommendation! I can't tell if it's quite equivalent to the use of this pattern though: it seems closer to a replacement for Flask's `req.args` call in [the next line](https://github.com/uwcirg/jwt-proxy/blob/9d5a155c60c03c946f47f4c4c0cc78ec24edf10b/jwt_proxy/api.py#L22) that parses the query params from the url, but if there's a better way to validate the content of a specific query param value there I'm all ears.

All that being said, I'll clean up this pattern - I'd hedged it a bit since I didn't know exactly what was url-encoded and what wasn't, but I've found that the query param values are always url-decoded by the `req.args` call here, and the check later against the identifier params in a POSTed resource's references will always come through url-encoded.